### PR TITLE
[FIRRTL] Fix LowerTypes for non-NLA hierpaths

### PR DESCRIPTION
**Pull request by seldridge**
Fix a bug in LowerTypes where it would only update hierarchical paths
that are involved in a non-local annotation (NLAs).  This is done in
preparation for more widespread use of hierarchical paths (e.g., by
sv::VerbatimOp) and in preparation for enabling hierarchical paths to
end at the module and not on the final operation.

It follows that this changes the lowering for a hierarchical path
targeting a component/port to generate one new hierarchical path _for
every_ ground type regardless of whether or not the hierarchical path is
used by a non-local annotation.  Previously, new hierarchical paths
would only be generated when used by non-local annotations.

This works by simplifying the way that annotations are updated in
LowerTypes.  Previously, this would update hierarchical paths when
examining annotations and keeping track of an "old symbol" to "new
symbol" mapping.  This commit changes the tracking to track "old
hw::InnerRefAttr" to "new Vector<AnnoTarget>".  This more closely
matches the lowering procedure (as LowerTypes is fundamentally a mapping
of one aggregate type to one or more ground types and results in a nice
simplification of the existing logic.

Signed-off-by: Schuyler Eldridge <schuyler.eldridge@sifive.com>

Example from the added test:

```mlir
firrtl.circuit "PathsWithoutNLAsAreStillUpdated" {
  firrtl.hierpath @path_port [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@a]
  firrtl.hierpath @path_wire [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@b]

  firrtl.module @Foo(in %a : !firrtl.bundle<a : uint<1>, b : bundle<c : uint<2>>> sym @a) {
    %b = firrtl.wire sym @b {annotations = [{class = "hello"}]} : !firrtl.bundle<a : uint<1>, b : bundle<c : uint<2>>>
  }
  firrtl.module @PathsWithoutNLAsAreStillUpdated() {
    %foo_a = firrtl.instance foo sym @foo @Foo(in a : !firrtl.bundle<a : uint<1>, b : bundle<c : uint<2>>>)
  }
}
```

Running this through `LowerTypes` produces the following debugging information:

```
===- Running LowerTypes Pass ------------------------------------------------===
Recording Inner Symbol Renames:
  - Module: @Foo
    - @b: [@b_a, @b_b_c]
    - @a: [@a_a, @a_b_c]
Updating hierarhical paths:
  - original: firrtl.hierpath @path_port [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@a]
    replacements:
      - firrtl.hierpath @path_port [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@a_a]
      - firrtl.hierpath @path_port_0 [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@a_b_c]
  - original: firrtl.hierpath @path_wire [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@b]
    replacements:
      - firrtl.hierpath @path_wire [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@b_a]
      - firrtl.hierpath @path_wire_0 [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@b_b_c]
```

And the MLIR output:

```mlir
firrtl.circuit "PathsWithoutNLAsAreStillUpdated"  {
  firrtl.hierpath @path_port [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@a_a]
  firrtl.hierpath @path_port_0 [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@a_b_c]
  firrtl.hierpath @path_wire [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@b_a]
  firrtl.hierpath @path_wire_0 [@PathsWithoutNLAsAreStillUpdated::@foo, @Foo::@b_b_c]
  firrtl.module @Foo(in %a_a: !firrtl.uint<1> sym @a_a, in %a_b_c: !firrtl.uint<2> sym @a_b_c) {
    %b_a = firrtl.wire sym @b_a  {annotations = [{class = "hello"}]} : !firrtl.uint<1>
    %b_b_c = firrtl.wire sym @b_b_c  {annotations = [{class = "hello"}]} : !firrtl.uint<2>
  }
  firrtl.module @PathsWithoutNLAsAreStillUpdated() {
    %foo_a_a, %foo_a_b_c = firrtl.instance foo sym @foo  @Foo(in a_a: !firrtl.uint<1>, in a_b_c: !firrtl.uint<2>)
  }
}
```

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:386]: This is dead code after the changes here. All the modifications to `filterAnnotations` are due to the implications of removing this.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:658]: The mechanics of how the port lowering works is that a module port list consisting of all old and new ports are created. After this, old ports are removed. This necessitates getting the final port index _after dead ports are deleted_ into `lowerArg` and `addArg` in order to populate `innerRefRenames` for ports. I would have preferred to not pass in another numeric argument, but the refactoring to get the code to work by maintaining separate old and new port lists proved tedious.

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (seldridge) [inline test/Dialect/FIRRTL/lower-types.mlir:1057]: This shows the main difference created by this PR. You get one hierarchical path _for every_ ground type as opposed to one hierarchical path for each NLA. This is a more restrictive, but a more strictly correct implementation of lowering.

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (prithayan) [inline test/Dialect/FIRRTL/lower-types.mlir:1049]: Isn't this invalid NLA ? `@NLALowering::@b_valid` doesn't have any annotation with `circt.nonlocal=@nla_b` . 
Maybe Out verifier should complain, but is ignoring this somehow.

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (prithayan) [inline test/Dialect/FIRRTL/lower-types.mlir:1062]: We should get rid of the breadcrumbs !

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (prithayan) [reply test/Dialect/FIRRTL/lower-types.mlir:1057]: Is that correct, shoud we get a `HierPathOp` for a ground type that does not have a nonlocal annotation ? The `HierPathOp` seems to be invalid in that case !

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (prithayan) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:563]: Since the rename map is used only for `HierPathOp`, what if we check for `circt.nonlocal` annotation here and insert the lowered field into the rename map only if the ground type has a nonlocal annotation.

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (seldridge) [reply test/Dialect/FIRRTL/lower-types.mlir:1049]: This is what this is trying to fix. The issue is that it's entirely legal to have a hierarchical path that is not used by an NLA, e.g., it could only be used by an `sv::VerbatimOp`.

Comment (prithayan) [reply test/Dialect/FIRRTL/lower-types.mlir:1049]: Ah, that makes sense. So, this should make it legal for any user.

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (seldridge) [reply test/Dialect/FIRRTL/lower-types.mlir]: This is fair. I had not changed it to prevent churn in the test, but this test is changed sufficiently that I think it's okay.

**Review by prithayan — APPROVED** [review]
LGTM.  Great that this cleans up the NLA handling mess in LowerTypes .

**Comment by seldridge** [comment]
Thanks for the review, @prithayan! I saw once crash on an internal design and will fix whatever is going on there before I land this.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (prithayan) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:1314]: It might be better if the update to the NLATable is done outside the loop. We can store a list of the `HierPathOp`s that need to be deleted and remove them outside the loop. This might be the cause of the crash you are seeing.

Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:1314]: Maybe your current implementation is the correct approach. You are already copying the result of the `lookup` on line 1288.
There are issues  in updating the `NLATable` while iterating over the result of a query `lookup`. Because, it returns a reference to the entry. 
If your crash is still present, try to valgrind it !

**Comment by seldridge** [comment]
There's a problem with this in that it doesn't work for aggregates with more than one level of nesting. E.g., the following fails verification because the rename is recorded for `@w_a -> @w_a_a` due to no rename being recorded for the first expansion:

```scala
circuit Foo: %[[
  {
    "class": "hello",
    "target": "~Foo|Foo/bar:Bar>w"
  }
]]
  module Bar:
    input a: {a: {b: UInt<1>}}
    output b: {a: {b: UInt<1>}}

    wire w: {a: {b: UInt<1>}}
    w <= a
    b <= w

  module Foo:
    input a: {a: {b: UInt<1>}}
    output b: {a: {b: UInt<1>}}

    inst bar of Bar
    bar.a <= a
    b <= bar.b
```

I'm working on a fix.

**Comment by seldridge** [comment]
I have a draft fix of this with a ton of debug code still in there. Wait until I pull the debug code out for a re-review. The change is to use a disjoint set (`EquivalenceClass`) to keep track of all the intermediary renames so that we can always dereference any `hw::InnerRefAttr` back to the original NLA symbol (leader).

A consequence of this is that it blows up verification time as this generates a gargantuan number of NLAs. There don't appear to be any implications for `--verify-each=false`. 

It may be better to stage this as part of the larger NLA refactor. I.e., this patch should be performant if we don't have symbols littered all through the design on larger aggregates. Alternatively, this motivates faster hierarchical path verification that is shared amongst multiple hierarchical path operations.

**Comment by seldridge** [comment]
Okay, final update and ready for re-review. In addition to the use of the disjoint set to track the original symbol, this is cleaned up in the following ways:

- The disjoint sets are per-module and only used during module lowering. They do not exit the parallel loop and are only used to keep `innerRefRenames` up-to-date.
- The update of the hierarchical paths now iterates directly over all hierarchical path ops as opposed to iterating over  `innerRefRenames` and filtering this based on NLATable usage.

**Comment by seldridge** [comment]
This is exposing pretty extreme hierarchical path-related performance issues in `PrefixModules` and `ModuleInliner`. There is also some crash in `GrandCentralTaps` related to an NLA not being associated with a hierarchical path. The performance issues are likely a blocker to merging this and the latter needs to be fixed (but also motivates annotation verification of some kind).

**Comment by prithayan** [comment]
I think I agree, it makes sense, to wait out for the `HierPathOp` to end on a module, before merging this PR.
Right now, as you observed, this will create a lot of spurious `HierPathOp`s, as there are no other users other than the nonlocal annotations.
Another change that should help this is per field symbol, then it should be much more direct.

**Comment by seldridge** [comment]
Agree.

Update: the crash is fixed. This was due to a reworking of the iteration logic overwriting all NLAs as opposed to only the ones it was sensitive to. I switched back to the old iteration order. 

This pass is fast, but exposes some really bad behavior in the aforementioned modules and in hierarchical path verification. This will need to sit until the other PRs are in place.

**Comment by seldridge** [comment]
The PR merged into this (#3300) applies changes to the default NLA format to not generate crazy numbers of hierarchical paths. This has the benefit of fixing the performance regression inherent in this PR in the inliner and prefix modules. This PR is unblocked.

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (youngar) [inline lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:151]: Can something like this work?
```
  for (size_t i = 0, e = nlas.size() - 1; i < e; ++i) {
    auto &nla = nlas[i];
    // Assumption: Symbol name = Operation name.
    auto module = std::get<1>(nla);
    auto inst = std::get<2>(nla);
    insts.push_back(hw::InnerRefAttr::get(StringAttr::get(ctxt, module),
                                          StringAttr::get(ctxt, inst)));
  }
  insts.push_back(FlatSymbolRefAttr::get(ctxt, module));
 ```

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (prithayan) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:1392]: symTbl->erase(path, symTbl)

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (prithayan) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:529]: I guess, it need not be a non-local DontTouchAnnotation, `needsSym=true` can be triggered even if a local `DontTouchAnnotation` is found!

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (prithayan) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:603]: Just an observation, there is nothing to record in the `innerRefRenames` if the original op didn't have an inner sym, but the lowered op has a symbol(when `needsSym=true`.

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:151]: Yeah, this looks better. 👍

**Comment by prithayan** [comment]
Went over `LowerTypes`, changes look good !! `EquivalenceClass` idea is pretty neat.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:529]: There is an assumption that any local `DontTouchAnnotation` is already represented as a symbol which this comment is assuming. I'll update this to indicate that this comment assumes that.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:603]: That should be expected and safe. If the original op didn't have a symbol, but gets one after `LowerTypes` then there can, by definition, be no valid rename.

File: lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp:151]: I'd be careful about `size_t... = size() - 1` re:wraparound, although looks like `nlas` being non-empty is already a precondition.  But when that goes wrong it's not always as easy to debug as other sorts of failures.

Anyway, since the index is no longer needed, consider replacing the `SmallVectorImpl<> &` with an `ArrayRef<>` (`MutableArrayRef` if this was written to) and you can simplify a bit more if you want:

```c++
static FlatSymbolRefAttr
buildNLA(CircuitOp circuit, size_t nlaSuffix,
         ArrayRef<std::tuple<std::string, StringRef, StringRef>> nlas) {
  OpBuilder b(circuit.getBodyRegion());
  MLIRContext *ctxt = circuit.getContext();
  SmallVector<Attribute> insts;
  for (auto &[module, inst, _] : nlas.drop_back()) {
    // Assumption: Symbol name = Operation name.
    insts.push_back(hw::InnerRefAttr::get(StringAttr::get(ctxt, module),
                                          StringAttr::get(ctxt, inst)));
  }
  insts.push_back(FlatSymbolRefAttr::get(ctxt, std::get<1>(nlas.back())));

```

**Comment by seldridge** [comment]
> Went over LowerTypes, changes look good !! EquivalenceClass idea is pretty neat.

Thanks for the additional review! Yeah, `EquivalenceClass` is super useful. I'm surprised that the only other usage is in `InferResets`.

**Comment by seldridge** [comment]
Opting to do a rare rebase-and-merge because these are very separate commits (and I made sure they pass regressions).

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp]: I thought even local `DontTouchAnnotation` on subfields are still preserved as `SubAnno` on the field, and not lowered to symbols!

**Pull request merged**
